File: 03_Slow_Fast_Pointers/05_happy_number/solution.py
def square(n):
    s = 0
    while n > 0:
        d = n % 10
        s += d * d
        n = n // 10
    return s

class Solution:
    def isHappy(self, n: int) -> bool:
        slow = fast = n
        while fast != 1:
            slow = square(slow)
            fast = square(square(fast))
            if slow == fast and slow != 1:
                return False
        return True

File: 03_Slow_Fast_Pointers/05_happy_number/solution.cpp
int square(int n){
    int sum = 0;
    while (n > 0){
        int digit = n % 10;
        sum += digit * digit;
        n /= 10;
    }
    return sum;
}

class Solution {
public:
    bool isHappy(int n) {
        int slow = n, fast = n;

        while (true){
            slow = square(slow);
            fast = square(square(fast));

            if (fast == 1) return true;

            if (slow == fast) return false;
        }
    }
};